presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java

```diff
@@ -164,29 +170,95 @@ public OrcSelectiveRecordReader(
             this.constantValues[zeroBasedIndices.get(entry.getKey())] = entry.getValue();
         }
 
+        Map<Integer, Map<Subfield, TupleDomainFilter>> filtersWithNewIndices = filters
+                .entrySet()
+                .stream()
+                .collect(toImmutableMap(entry -> zeroBasedIndices.get(entry.getKey()), Map.Entry::getValue ))
+                ;
+
         // Initial order of stream readers is:
-        //  - readers with simple filters
+        //  - readers with integer equality
+        //  - readers with integer range / multivalues / inequality
+        //  - readers with filters
         //  - followed by readers for columns that provide input to filter functions
         //  - followed by readers for columns that doesn't have any filtering
-        streamReaderOrder = orderStreamReaders(columnTypes.keySet().stream().filter(index -> this.constantValues[index] == null).collect(toImmutableSet()), columnsWithFilters, filterFunctionInputs);
+        streamReaderOrder = orderStreamReaders(columnTypes.keySet().stream().filter(index -> this.constantValues[index] == null).collect(toImmutableSet()), filtersWithNewIndices, filterFunctionInputs);
     }
 
-    private static int[] orderStreamReaders(Collection<Integer> columnIndices, Set<Integer> columnsWithFilters, Set<Integer> filterFunctionInputs)
+    private static int scoreFilter(Map<Subfield, TupleDomainFilter> filters){
+        final int lowestScore = 10;
+        final int lowScore = 50;
+        final int mediumScore = 100;
+        final int maxScore = 1000;
+
+        checkArgument(!filters.isEmpty());
+
+        if (filters.size() > 1){
+            // This can only be true if this column is a complex type.
```

Comment:
I'd shorten this comment and make it fit on one line: `// Complex type column. Complex types are expensive!`

Reply:
Okay!